services/github/src/handlers.js

```diff
@@ -220,12 +225,32 @@ class Handlers {
     }
   }
 
+  commentKey(idents) {
+    return crypto
+      .createHash('md5')
+      .update(stringify(idents))
+      .digest('hex');
+  }
+
+  isDuplicateComment(...idents) {
+    return _.indexOf(this.commentHashCache, this.commentKey(idents)) !== -1;
```

Comment:
My brain is in optimization mode. It probably doesn't matter. And yet, I need to comment so I can move on...

It seems inefficient to iterate through up to 1,000 items in the (hopefully default) case of adding a new comment. A data structure like a [Set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) would be a better match for this part. But, it doesn't allow removing the oldest item.

I guess what we want is an ordered set, so we can quickly test for inclusion, and also remove the last item. I've seen this implemented as a list, to remember insertion order, paired with a set, for quick lookups. To rotate the cache, you pop the end of the list, and remove the related element from the set.

This should work as a proof of concept, and could be extended to an ordered set later.

Reply:
We use an LRU cache library a few other places.

Reply:
Yeah agreed. This just stole the approach from how we do it in tc-notify which could also be improved with the library I expect. Just going to land this as-is now though and put off a smarter approach for later.